<h4 style="color:#1a96f6;"> <center> <b>

MIT License

Copyright (c) 2024 Felipe Braz da Silva e Pedro Wilson Silva de Sousa

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
</b> </center> </h4>

<div class="alert alert-info">
    
<h1 style="color:#1a96f6;"> <center> <b>
Projeto Final- Disciplina Técnicas de Prototipagem
</b> </center> </h1>
    
<h3 style="color:#1a96f6;"> <center> <b>
Projeto ECO
</b></center> </h3>
    
<h3 style="color:#1a96f6;"> <center> <b>
Prof. Dr. Moacy Pereira da Silva
</b></center> </h3>

<h3 style="color:#1a96f6;"> <center> <b>
Discentes: 

Felipe Braz da Silva

Pedro Wilson Silva de Sousa
</b></center> </h3>

<h3 style="color:#1a96f6;"> <center> <b>
Data: 16/19/2024
</b></center> </h3>


</div>

# Projeto ECO

#### *Objetivo do módulo*: Treinar um modelo de rede neural convulacional para identificação de materiais recicláveis.



##### Considerações:

A príncipio, o projeto visa identificar os materiais vidro e metal. Posteriormente, em pesquisas futuras consideramos extender a identificação a outros materiais comuns no processo de coleta seletiva. 

### Passo 1: Carregamento das imagens e divisão em treino e teste

As imagens foram adquiridas a partir do dataset Recyclable Solid Wast disponível na plataforma kaggle.

link para dataset: https://www.kaggle.com/datasets/hseyinsaidkoca/recyclable-solid-waste-dataset-on-5-background-co

#### 1.1 A ferramenta ImageDataGenerator 

A ferramenta ImageDataGenerator foi utilizada para dividir os dados em treinamento e teste. Utilizamos a proporção de 80% para treinamento e 20% para teste escolhidos de forma aleatória, padrão bem aceito em projetos de Machine Learning e Deep Learning.

Conjunto de treino: 424 imagens

Conjunto de teste: 161 imagens

In [4]:
from tensorflow.keras.preprocessing.image import ImageDataGenerator

dataset_path = 'C:/Users/felip/PycharmProjects/Notebooks IA/dataset'

# Diretórios de treinamento e teste
train_dir = dataset_path + '/train'
test_dir = dataset_path + '/test'

# Inicializar ImageDataGenerator para normalização
train_datagen = ImageDataGenerator(rescale=1./255)
test_datagen = ImageDataGenerator(rescale=1./255)

# Carregar o conjunto de treinamento
train_generator = train_datagen.flow_from_directory(
    train_dir,
    target_size=(150, 150), 
    batch_size=32,
    class_mode='categorical' 
)

# Carregar o conjunto de teste
test_generator = test_datagen.flow_from_directory(
    test_dir,
    target_size=(150, 150),
    batch_size=32,
    class_mode='categorical'
)


Found 424 images belonging to 3 classes.
Found 161 images belonging to 3 classes.


In [3]:
print("Train generator: {} samples found.".format(train_generator.samples))
print("Test generator: {} samples found.".format(test_generator.samples))


Train generator: 424 samples found.
Test generator: 161 samples found.


### Passo 2: Construção do modelo de rede neural Convulacional

Utilizando como referência as ferramentas de modelagem de redes neurais mais atuais foi escolhida a utilização da biblioteca tensorflow com o módulo Keras.


*Modelo 1*: AlexNet, modelo de rede neural convulacional aplicado em Deep learning desenvolvido por Alex Krizhevsky, Ilya Sutskever e Geoff Hinton. Conhecido por vencer o ImageNet Large Scale Visual Recognition Challenge (ILSVRC) em 2012.

Mais informações:

https://vitalflux.com/different-types-of-cnn-architectures-explained-examples/

https://medium.com/@siddheshb008/alexnet-architecture-explained-b6240c528bd5]


*Modelo 2*: Modelo aleatório

In [18]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout

model = Sequential(
    [
        Conv2D(96, kernel_size=(11, 11), strides=4,
               padding='valid', activation='relu',
               input_shape=(150, 150, 3),
               kernel_initializer='he_normal'),
    
        MaxPooling2D(pool_size=(3, 3), strides=(2, 2),
                     padding='valid', data_format=None),

        Conv2D(256, kernel_size=(5, 5), strides=1,
               padding='same', activation='relu',
               kernel_initializer='he_normal'),
    
        MaxPooling2D(pool_size=(3, 3), strides=(2, 2),
                     padding='valid', data_format=None),
        
        Conv2D(384, kernel_size=(3, 3), strides=1,
               padding='same', activation='relu',
               kernel_initializer='he_normal'),
    
        Conv2D(384, kernel_size=(3, 3), strides=1,
               padding='same', activation='relu',
               kernel_initializer='he_normal'),

        Conv2D(256, kernel_size=(3, 3), strides=1,
               padding='same', activation='relu',
               kernel_initializer='he_normal'),

        MaxPooling2D(pool_size=(3, 3), strides=(2, 2),
                     padding='valid', data_format=None),

        Flatten(),
    
        Dense(4096, activation='relu'),
        Dense(4096, activation='relu'),
        Dense(1000, activation='relu'),
        Dense(3, activation='softmax')
    ]
)

model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])
model.summary()


Model: "sequential_4"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_18 (Conv2D)              │ (None, 35, 35, 96)     │        34,944 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_14 (MaxPooling2D) │ (None, 17, 17, 96)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_19 (Conv2D)              │ (None, 17, 17, 256)    │       614,656 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_15 (MaxPooling2D) │ (None, 8, 8, 256)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_20 (Conv2D)              │ (None, 8, 8, 384)      │       885,120 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_21 (Conv2D)              │ (None, 8, 8, 384)      │     1,327,488 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_22 (Conv2D)              │ (None, 8, 8, 256)      │       884,992 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_16 (MaxPooling2D) │ (None, 3, 3, 256)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_4 (Flatten)             │ (None, 2304)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_12 (Dense)                │ (None, 4096)           │     9,441,280 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_13 (Dense)                │ (None, 4096)           │    16,781,312 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_14 (Dense)                │ (None, 1000)           │     4,097,000 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_15 (Dense)                │ (None, 3)              │         3,003 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 34,069,795 (129.97 MB)

 Trainable params: 34,069,795 (129.97 MB)

 Non-trainable params: 0 (0.00 B)

In [28]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout

model_2 = Sequential(
    [
        Conv2D(32, (3, 3), activation='relu', input_shape=(150, 150, 3)),
        MaxPooling2D(pool_size=(2, 2)),
        
        Conv2D(64, (3, 3), activation='relu'),
        MaxPooling2D(pool_size=(2, 2)),
        
        Conv2D(128, (3, 3), activation='relu'),
        MaxPooling2D(pool_size=(2, 2)),
        
        Conv2D(128, (3, 3), activation='relu'),
        MaxPooling2D(pool_size=(2, 2)),
        
        Flatten(),
    
        Dense(512, activation='relu'),
    
        Dropout(0.5),
    
        Dense(3, activation='softmax')    
    ]

)

model_2.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

model_2.summary()


Model: "sequential_8"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_35 (Conv2D)              │ (None, 148, 148, 32)   │           896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_29 (MaxPooling2D) │ (None, 74, 74, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_36 (Conv2D)              │ (None, 72, 72, 64)     │        18,496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_30 (MaxPooling2D) │ (None, 36, 36, 64)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_37 (Conv2D)              │ (None, 34, 34, 128)    │        73,856 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_31 (MaxPooling2D) │ (None, 17, 17, 128)    │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_38 (Conv2D)              │ (None, 15, 15, 128)    │       147,584 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_32 (MaxPooling2D) │ (None, 7, 7, 128)      │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_8 (Flatten)             │ (None, 6272)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_22 (Dense)                │ (None, 512)            │     3,211,776 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_5 (Dropout)             │ (None, 512)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_23 (Dense)                │ (None, 3)              │         1,539 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 3,454,147 (13.18 MB)

 Trainable params: 3,454,147 (13.18 MB)

 Non-trainable params: 0 (0.00 B)

### Passo 3: Treinamento do modelo de rede neural Convulacional

In [29]:
# Treinando o modelo 1: AlexNet
history_model_1 = model.fit(
    train_generator,
    steps_per_epoch=train_generator.samples // train_generator.batch_size,
    epochs=25,  
    validation_data=test_generator,
    validation_steps=test_generator.samples // test_generator.batch_size
)


Epoch 1/25
13/13 ━━━━━━━━━━━━━━━━━━━━ 4s 271ms/step - accuracy: 0.8777 - loss: 0.3071 - val_accuracy: 0.9000 - val_loss: 0.2955
Epoch 2/25
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9375 - loss: 0.2547 - val_accuracy: 0.0000e+00 - val_loss: 2.6851
Epoch 3/25
13/13 ━━━━━━━━━━━━━━━━━━━━ 4s 256ms/step - accuracy: 0.8771 - loss: 0.3746 - val_accuracy: 0.9062 - val_loss: 0.2830
Epoch 4/25
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.8125 - loss: 0.3871 - val_accuracy: 0.0000e+00 - val_loss: 0.9883
Epoch 5/25
13/13 ━━━━━━━━━━━━━━━━━━━━ 4s 250ms/step - accuracy: 0.8867 - loss: 0.3093 - val_accuracy: 0.8938 - val_loss: 0.2276
Epoch 6/25
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9062 - loss: 0.2448 - val_accuracy: 1.0000 - val_loss: 0.1784
Epoch 7/25
13/13 ━━━━━━━━━━━━━━━━━━━━ 4s 256ms/step - accuracy: 0.8825 - loss: 0.3020 - val_accuracy: 0.9438 - val_loss: 0.1922
Epoch 8/25
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.9062 - loss: 0.1843 - val_accuracy: 1.

In [30]:
# Treinando o modelo 2: Aleatório
history_model_2 = model_2.fit(
    train_generator,
    steps_per_epoch=train_generator.samples // train_generator.batch_size,
    epochs=25,  
    validation_data=test_generator,
    validation_steps=test_generator.samples // test_generator.batch_size
)

Epoch 1/25
13/13 ━━━━━━━━━━━━━━━━━━━━ 4s 173ms/step - accuracy: 0.3934 - loss: 1.1051 - val_accuracy: 0.4375 - val_loss: 1.0327
Epoch 2/25
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 3ms/step - accuracy: 0.4688 - loss: 1.0660 - val_accuracy: 1.0000 - val_loss: 0.8166
Epoch 3/25
13/13 ━━━━━━━━━━━━━━━━━━━━ 2s 156ms/step - accuracy: 0.4028 - loss: 1.0431 - val_accuracy: 0.7563 - val_loss: 0.8795
Epoch 4/25
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.6562 - loss: 0.9135 - val_accuracy: 1.0000 - val_loss: 0.4847
Epoch 5/25
13/13 ━━━━━━━━━━━━━━━━━━━━ 2s 166ms/step - accuracy: 0.6041 - loss: 0.8474 - val_accuracy: 0.8687 - val_loss: 0.3933
Epoch 6/25
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.8750 - loss: 0.3751 - val_accuracy: 1.0000 - val_loss: 0.0031
Epoch 7/25
13/13 ━━━━━━━━━━━━━━━━━━━━ 2s 160ms/step - accuracy: 0.7929 - loss: 0.4733 - val_accuracy: 0.8625 - val_loss: 0.3377
Epoch 8/25
13/13 ━━━━━━━━━━━━━━━━━━━━ 0s 2ms/step - accuracy: 0.7812 - loss: 0.4765 - val_accuracy: 1.0000 - v

### Passo 4: Avaliação do modelo de rede neural Convulacional

In [31]:
# Avaliando o modelo 1
test_loss, test_accuracy = model.evaluate(test_generator)
print("Test Accuracy: {:.2f}".format(test_accuracy * 100))


6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 69ms/step - accuracy: 0.9612 - loss: 0.1399
Test Accuracy: 95.65


In [32]:
# Avaliando o modelo 2
test_loss, test_accuracy = model_2.evaluate(test_generator)
print("Test Accuracy: {:.2f}".format(test_accuracy * 100))


6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 52ms/step - accuracy: 0.9305 - loss: 0.2293
Test Accuracy: 93.79


De acordo com os resultados obtidos foi escolhido o modelo AlexNet com desempenho de 95.6 % de Acurácia. Para o intuito do projeto podemos considerar um bom desempenho, não deixando de considerar a possibilidade de realizar pesquisas futuras para refinamento e escolha do modelo na rede neural convulacional adequado e execução de baterias de teste para devida validação do modelo.

**Observação**: o modelo AlexNet foi projetado para ser usado com conjuntos de dados de imagens em larga escala, mas para testar o comportamento obtido com nosso dataset decidimos testar.

### Passo 5: Salvar o modelo

In [34]:
# Salvando o modelo
model.save('alexNet.h5')
model_2.save('cnn1.h5')